numpy/_core/src/multiarray/ctors.c

```diff
@@ -39,6 +39,13 @@
 
 #include "umathmodule.h"
 
+
+NPY_NO_EXPORT const char *npy_no_copy_err_msg = (
+        "Unable to avoid copy while creating an array as requested.\n"
+        "If using `np.array(obj, copy=False)` use `np.asarray(obj)` "
+        "or `copy=None` to allow NumPy to make the copy.\n"
+        "This changed in NumPy 2. The suggested fix works on all versions.");
```

Comment:
This is not quite right, `copy=None` does not work on numpy 1.x:
```
In [2]: np.array(1, copy=None)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [2], line 1
----> 1 np.array(1, copy=None)

ValueError: NoneType copy mode not allowed.
```

Reply:
🤯, I am *really* surprised that this doesn't work!  @mtsokol what would be the suggestion then, just `np.asarray`?  It means that forwarding a kwarg like:
```
def myfunc(arr, copy=None):
    return np.array(arr, copy=copy)
```
has no trivial cross-NumPy solution!?

Reply:
Right, for NumPy 1.x `np.asarray(arr)` is necessary. In Pandas we did (https://github.com/pandas-dev/pandas/pull/57172):

```
copy_false = None if np_version_gt2 else False
```

Reply:
I removed the `copy=None` suggestion and added a half sentence for now.  But maybe we need to mention that you may need `copy_if_needed = None if np_version_gt2 else False` to do this right?

Reply:
Indeed, in astropy we do the same: `COPY_IF_NEEDED = False if NUMPY_LT_2_0 else None`

Since that was just copied from scipy, I think it makes sense to include it in the error message...

Reply:
I am wondering if it makes more sense to put a link to the migration guide when it comes beyond pointing to `np.asarray()`?

Reply:
Yes, putting a link makes sense!

Reply:
Tweaked it a bit, the link is unfortunately long, but probably just shouldn't worry about it :).  Better too verbose than confusing!